src/libraries/Microsoft.Extensions.DependencyInjection/tests/DI.Tests/ServiceProviderContainerTests.cs

```diff
@@ -406,6 +406,219 @@ public void Dispose()
             }
         }
 
+        [ThreadStatic]
+        public static int ThreadId;
+
+        [Fact]
+        public async Task GetRequiredService_UsesSingletonAndLazyLocks_NoDeadlock()
+        {
+            // Thread 1: Thing1 (transient) -> Thing0 (singleton)
+            // Thread 2: Thing2 (singleton) -> Thing1 (transient) -> Thing0 (singleton)
+
+            // 1. Thread 1 resolves the Thing1 which is a transient service
+            // 2. In parallel, Thread 2 resolves Thing2 which is a singleton
+            // 3. Thread 1 enters the factory callback for Thing1 and takes the lazy lock
+            // 4. Thread 2 takes callsite for Thing2 as a singleton lock when it resolves Thing2
+            // 5. Thread 2 enters the factory callback for Thing1 and waits on the lazy lock
+            // 6. Thread 1 calls GetRequiredService<Thing0> on the service provider, takes callsite for Thing0 causing no deadlock
+            // (rather than taking the locks that are already taken - either the lazy lock or the Thing2 callsite lock)
+
+            Thing0 thing0 = null;
+            Thing1 thing1 = null;
+            Thing2 thing2 = null;
+            IServiceProvider sp = null;
+
+            // Arrange
+            var services = new ServiceCollection();
+
+            var lazy = new Lazy<Thing1>(() =>
+            {
+                thing0 = sp.GetRequiredService<Thing0>();
+                return new Thing1(thing0);
+            });
+
+            services.AddSingleton<Thing0>();
+            services.AddTransient(sp =>
+            {
+                if (ThreadId == 1)
+                {
+                    Action1();
+                }
+                else
+                {
+                    // Let Thread 1 over take Thread 2
+                    Action2();
+                }
+
+                return lazy.Value;
+            });
+            services.AddSingleton<Thing2>();
+
+            sp = services.BuildServiceProvider();
+
+            var t1 = Task.Run(() =>
+            {
+                ThreadId = 1;
+                using var scope1 = sp.CreateScope();
+                thing1 = scope1.ServiceProvider.GetRequiredService<Thing1>();
+            });
+
+            var t2 = Task.Run(() =>
+            {
+                ThreadId = 2;
+                using var scope2 = sp.CreateScope();
+                thing2 = scope2.ServiceProvider.GetRequiredService<Thing2>();
+            });
+
+            // Act
+            await t1;
+            await t2;
+
+            // Assert
+            Assert.NotNull(thing0);
+            Assert.NotNull(thing1);
+            Assert.NotNull(thing2);
+        }
+
+        private ManualResetEvent _manualResetEvent = new ManualResetEvent(false);
```

Comment:
Instead of an instance member, can you create new ones of these in each test? Currently you aren't resetting the event in between tests, so I'm not sure the 2nd test is testing the scenario properly.

Reply:
From https://xunit.net/docs/shared-context#constructor, 

> xUnit.net creates a new instance of the test class for every test that is run, so any code which is placed into the constructor of the test class will be run for every single test.